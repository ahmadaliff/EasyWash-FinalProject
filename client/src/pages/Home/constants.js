export const GET_MERCHANTS = 'home/GET_MERCHANT';
export const SET_MERCHANTS = 'home/SET_MERCHANT';
export const RESET_MERCHANTS = 'home/RESET_MERCHANT';
