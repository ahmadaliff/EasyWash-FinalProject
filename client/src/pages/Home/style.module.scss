@import '@styles/base';
@import url('https://fonts.googleapis.com/css2?family=Lobster&display=swap');

.mainWrap {
  --image-size-sm: 10rem;
  --image-size-md: 12rem;
  @include directionY;
  padding: 0 0 2rem 0;
  min-height: calc(100vh - 3.55rem);
  align-items: center;
  box-sizing: border-box;
  position: relative;
  * {
    box-sizing: border-box;
  }
  .head {
    @include directionY;
    align-items: center;
    gap: 0;
    @include apply-at-sm-size {
      @include directionX;
      gap: 2rem;
      justify-content: center;
    }
    width: 100%;
    .ilustration {
      max-width: 20rem;
      @include apply-at-md-size {
        max-width: 40vw;
      }
    }
    .headerText {
      @include directionX;
      align-items: center;
      font-size: 2.5rem;
      max-width: 15rem;
      @include apply-at-md-size {
        font-size: 3rem;
        max-width: 15rem;
      }
      font-family: 'Lobster', cursive;
    }
  }
  .content {
    display: grid;
    grid-template-columns: repeat(2, minmax(10rem, 10rem));
    @include apply-at-sm-size {
      grid-template-columns: repeat(2, minmax(15rem, 1fr));
    }
    @include apply-at-lg-size {
      grid-template-columns: repeat(3, minmax(15rem, 1fr));
    }
    @include apply-at-xl-size {
      grid-template-columns: repeat(4, minmax(15rem, 1fr));
    }
    gap: 1rem;
    padding-top: 2rem;
  }
  .noData {
    @include directionX;
    align-items: center;
    min-height: 40vh;
  }
}
