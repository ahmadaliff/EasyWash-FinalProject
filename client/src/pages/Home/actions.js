import { GET_MERCHANTS, RESET_MERCHANTS, SET_MERCHANTS } from './constants';

export const actionGetMerchants = () => ({
  type: GET_MERCHANTS,
});

export const actionSetMerchants = (data) => ({
  type: SET_MERCHANTS,
  data,
});

export const actionResetMerchants = () => ({
  type: RESET_MERCHANTS,
});
