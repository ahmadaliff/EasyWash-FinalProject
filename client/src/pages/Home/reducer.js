import { produce } from 'immer';
import { RESET_MERCHANTS, SET_MERCHANTS } from './constants';

export const initialState = {
  merchants: null,
};

export const storedKey = ['merchants'];

export const homeReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_MERCHANTS:
        draft.merchants = action.data;
        break;
      case RESET_MERCHANTS:
        return initialState;
    }
  });

export default homeReducer;
