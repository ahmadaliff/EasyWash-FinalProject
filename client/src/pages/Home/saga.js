import { call, put, takeLatest } from 'redux-saga/effects';
import { setLoading, showPopup } from '@containers/App/actions';
import toast from 'react-hot-toast';
import { apiGetMerchants } from '@domain/api';
import intlHelper from '@utils/intlHelper';
import { GET_MERCHANTS } from './constants';
import { actionSetMerchants } from './actions';

const getUserLocation = () =>
  new Promise((resolve, reject) => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(
        (location) => {
          const { latitude, longitude } = location.coords;
          const loc = { location: JSON.stringify({ lat: latitude, lng: longitude }) };
          resolve(loc);
        },
        (error) => reject(error)
      );
    }
  });

function* sagaGetMerchants() {
  yield put(setLoading(true));
  try {
    const location = yield call(getUserLocation);
    const response = yield call(apiGetMerchants, location);
    yield put(actionSetMerchants(response.data));
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404 || error?.response?.status === 401) {
      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

export default function* homeSaga() {
  yield takeLatest(GET_MERCHANTS, sagaGetMerchants);
}
