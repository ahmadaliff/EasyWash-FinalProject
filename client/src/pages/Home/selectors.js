import { createSelector } from 'reselect';
import { initialState } from '@pages/Home/reducer';

const selectHomeState = (state) => state.home || initialState;

export const selectMerchants = createSelector(selectHomeState, (state) => state.merchants);
