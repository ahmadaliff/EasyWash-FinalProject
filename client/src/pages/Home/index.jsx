import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import MerchantCard from '@components/MerchantCard';
import NoData from '@components/NoData';

import { Map, Search } from '@mui/icons-material';

import { selectMerchants } from '@pages/Home/selectors';
import { actionGetMerchants, actionResetMerchants } from '@pages/Home/actions';

import classes from '@pages/Home/style.module.scss';
import { IconButton } from '@mui/material';
import DialogMerchantsMap from '@components/DialogMerchantsMap';

const Home = ({ merchants }) => {
  const dispatch = useDispatch();
  const [filterSearch, setFilterSearch] = useState('');
  const [isMapOpen, setIsMapOpen] = useState(false);
  useEffect(() => {
    dispatch(actionGetMerchants());
    return () => {
      if (merchants) {
        dispatch(actionResetMerchants());
      }
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch]);

  return (
    <main className={classes.mainWrap}>
      <header className={classes.head}>
        <div className={classes.headerText}>
          <h2>
            <FormattedMessage id="app_tagline" />
          </h2>
          <div className={classes.searchInputWrap}>
            <Search className={classes.iconSearch} />
            <input className={classes.searchInput} onChange={(e) => setFilterSearch(e.target.value)} />
            <IconButton color="inherit" onClick={() => setIsMapOpen(true)}>
              <Map />
            </IconButton>
          </div>
        </div>
        <div className={classes.ilustrationWrap}>
          <img src="homePage.svg" className={classes.ilustration} alt="https://storyset.com" />
          <a href="https://storyset.com/web" className={classes.storySet}>
            Web illustrations by Storyset
          </a>
        </div>
      </header>
      <header className={classes.secondHeader}>
        <h3>
          <FormattedMessage id="app_looking_for_laundry" />
        </h3>
        <p>
          <FormattedMessage id="app_second_header_description" />
        </p>
      </header>
      <div className={classes.content}>
        {merchants
          ?.filter(({ name }) => name.toLowerCase().includes(filterSearch.toLowerCase()))
          .map((val, key) => (
            <MerchantCard merchant={val} key={key} />
          ))}
      </div>
      {merchants?.filter(({ name }) => name.toLowerCase().includes(filterSearch.toLowerCase()))?.length === 0 && (
        <NoData />
      )}
      <DialogMerchantsMap merchants={merchants} open={isMapOpen} handleClose={() => setIsMapOpen(false)} />
    </main>
  );
};

Home.propTypes = {
  merchants: PropTypes.array,
};

const mapStateToProps = createStructuredSelector({
  merchants: selectMerchants,
});

export default connect(mapStateToProps)(Home);
