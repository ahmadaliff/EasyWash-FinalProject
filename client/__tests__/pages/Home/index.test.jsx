import { render, fireEvent } from '@utils/testHelper';
import Home from '@pages/Home';

let wrapper;
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));
jest.mock('react-router-dom', () => ({
  useNavigate: () => jest.fn(),
}));

jest.mock('@pages/Home/selectors', () => ({
  selectMerchants: () => [
    {
      name: 'merchant',
      imagePath: null,
      distance: 5,
      description: 'description merchant',
      Favorits: [],
      userId: 2,
    },
  ],
}));

jest.mock('@components/DialogMerchantsMap', () =>
  jest.fn().mockImplementation(() => <div data-testid="dialog-merchants-map" />)
);

beforeEach(() => {
  wrapper = render(<Home />);
});

describe('Home Page', () => {
  test('renders', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('home-wrap')).toBeInTheDocument();
  });

  test('handles search input', () => {
    const { getByTestId } = wrapper;
    const searchInput = getByTestId('search-input');
    fireEvent.change(searchInput, { target: { value: 'search value' } });
    expect(searchInput.value).toBe('search value');
  });

  test('opens DialogMerchantsMap on map icon click', () => {
    const { getByTestId } = wrapper;
    fireEvent.click(getByTestId('dialog-merchants-map'));
  });
  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
