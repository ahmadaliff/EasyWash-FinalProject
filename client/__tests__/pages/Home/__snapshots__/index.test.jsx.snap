// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="mainWrap"
        data-testid="home-wrap"
      >
        <header
          class="head"
        >
          <div
            class="headerText"
          >
            <h2>
              Pengalaman mencuci Anda lebih mudah !!
            </h2>
            <div
              class="searchInputWrap"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium iconSearch css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <input
                class="searchInput"
                data-testid="search-input"
              />
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="MapIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m20.5 3-.16.03L15 5.1 9 3 3.36 4.9c-.21.07-.36.25-.36.48V20.5c0 .28.22.5.5.5l.16-.03L9 18.9l6 2.1 5.64-1.9c.21-.07.36-.25.36-.48V3.5c0-.28-.22-.5-.5-.5zM15 19l-6-2.11V5l6 2.11V19z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="ilustrationWrap"
          >
            <img
              alt="https://storyset.com"
              class="ilustration"
              src="homePage.svg"
            />
            <a
              class="storySet"
              href="https://storyset.com/web"
            >
              Web illustrations by Storyset
            </a>
          </div>
        </header>
        <header
          class="secondHeader"
        >
          <h3>
            Sedang mencari laundry?
          </h3>
          <p>
            Temukan laundry dengan layanan terbaik dan penawaran terbaik, kami tampilkan laundry dalam radius 3 km.
          </p>
        </header>
        <div
          class="content"
        >
          <div
            class="cardWrap"
            data-testid="card-wrap"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card css-bhp9pd-MuiPaper-root-MuiCard-root"
              data-testid="card-merchant"
            >
              <div
                class="MuiAvatar-root MuiAvatar-square MuiAvatar-colorDefault avatar css-18gxrea-MuiAvatar-root"
              >
                m
              </div>
              <div
                class="logoDelivery"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="LocalShippingIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 8h-3V4H3c-1.1 0-2 .9-2 2v11h2c0 1.66 1.34 3 3 3s3-1.34 3-3h6c0 1.66 1.34 3 3 3s3-1.34 3-3h2v-5l-3-4zM6 18.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm13.5-9 1.96 2.5H17V9.5h2.5zm-1.5 9c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                  />
                </svg>
                Gratis Ongkir
              </div>
              <h4
                class="headerText"
              >
                merchant
              </h4>
              <p
                class="distance"
              >
                5
                 
                Km
              </p>
              <p
                class="desc"
              >
                description merchant
              </p>
            </div>
          </div>
        </div>
        <div
          data-testid="dialog-merchants-map"
        />
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="mainWrap"
      data-testid="home-wrap"
    >
      <header
        class="head"
      >
        <div
          class="headerText"
        >
          <h2>
            Pengalaman mencuci Anda lebih mudah !!
          </h2>
          <div
            class="searchInputWrap"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium iconSearch css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
            <input
              class="searchInput"
              data-testid="search-input"
            />
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MapIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m20.5 3-.16.03L15 5.1 9 3 3.36 4.9c-.21.07-.36.25-.36.48V20.5c0 .28.22.5.5.5l.16-.03L9 18.9l6 2.1 5.64-1.9c.21-.07.36-.25.36-.48V3.5c0-.28-.22-.5-.5-.5zM15 19l-6-2.11V5l6 2.11V19z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="ilustrationWrap"
        >
          <img
            alt="https://storyset.com"
            class="ilustration"
            src="homePage.svg"
          />
          <a
            class="storySet"
            href="https://storyset.com/web"
          >
            Web illustrations by Storyset
          </a>
        </div>
      </header>
      <header
        class="secondHeader"
      >
        <h3>
          Sedang mencari laundry?
        </h3>
        <p>
          Temukan laundry dengan layanan terbaik dan penawaran terbaik, kami tampilkan laundry dalam radius 3 km.
        </p>
      </header>
      <div
        class="content"
      >
        <div
          class="cardWrap"
          data-testid="card-wrap"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card css-bhp9pd-MuiPaper-root-MuiCard-root"
            data-testid="card-merchant"
          >
            <div
              class="MuiAvatar-root MuiAvatar-square MuiAvatar-colorDefault avatar css-18gxrea-MuiAvatar-root"
            >
              m
            </div>
            <div
              class="logoDelivery"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LocalShippingIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 8h-3V4H3c-1.1 0-2 .9-2 2v11h2c0 1.66 1.34 3 3 3s3-1.34 3-3h6c0 1.66 1.34 3 3 3s3-1.34 3-3h2v-5l-3-4zM6 18.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm13.5-9 1.96 2.5H17V9.5h2.5zm-1.5 9c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                />
              </svg>
              Gratis Ongkir
            </div>
            <h4
              class="headerText"
            >
              merchant
            </h4>
            <p
              class="distance"
            >
              5
               
              Km
            </p>
            <p
              class="desc"
            >
              description merchant
            </p>
          </div>
        </div>
      </div>
      <div
        data-testid="dialog-merchants-map"
      />
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
